tripal/tripal
# Tv4g1 issue1843 publishing cardinality

# Bug Fix

### Issue #1843


### Tripal Version: 4.x

## Description



This PR solves the issue described in issue #1843 where field values derived from linked tables in Chado were only showing the first item and ignoring the others.  In actuality the problem was that the testing we did made it look like it was publishing the first item but in fact it was publishing the wrong thing but because record IDs were similar between the base record and linked record we didnt notice. For example.  If you add a project and link two contacts to it, then publishin the contacts on the project, it looked like it only added one, but it was really publishing a contact with the same record ID as the project not the contact. Since the first contact was what was desired it looked like it was working .

There were multiple issues that this PR solves in both ChadoStorage and TripalPublish:

1.  There was a bug introduced that prevented title tokens from being renamed. I fixed that with the help of @dsenalik and @laceysanderson 
2.  The `ChadoStorage::findValues()` function sets the field property values after it finds all of the fields.  One property that all fields have is the `record_id` which always points to the base record.  When we find values, we always include the fields in the values array passed in because we want to find any field items to publish.  This resulted in the `findValues` function setting the `record_id` for the first item for all fields regardless if any items were found.  This resulted in publishing of incomplete items. They didn't seem to show up on the pages, but they were in the Drupal tables. This was fixed by adding a new function named `ChadoStorage::isFieldValid()` that the `findValues()` uses to make sure a field has all of the required values before including it in what is returned.  I also add a `has_values` to the `ChadoRecords` array that is set if the `values` for the item are really set.  This allows the `isFieldValid()` to really be sure that an item has a set value or not (i.e. empty values are sometimes okay).   Also, the `findValues()` function was presetting the IDs and Links on linked tables and it should not have been doing that.  
3. The reason that we could publish all properties but not all contacts on a project was because the `ChadoRecords::findRecords()` function was not looking at all of the ancialliary tables.  There was a bug in that it exepcted that at least on of the columns in the linked table had to be used. This is the case in a property table where the `value` column is used, but in the case of a linking table, we don't need any other columns but the IDs.  So, once the problem in item 2 above was fixed, no `project_contact` records were being published at all.  I fixed this so the `findRecords()` function now will go through all linked tables.  This is fixed.
4. The `TripalPublish::insertFieldItems()` had a bug in that it wasn't properly cycling through the entities correctly. This is fixed.  I also fixed the way the number of published field items was reported.  

## Testing?



1. Load a brand new site.
2. Create an analysis
3. Create an organism
4. Create 2 or more projects
5. Create 2 or more contacts
6. Add a lot of records using the GFF loader and publish content from it.  This tests that normal publishing is still working
7. Link the contacts to projects with this SQL statement: `insert into project_contact (project_id, contact_id) values (1,2), (1,3), (2,2), (2,3);`
8. Now publish the `project`. It should add two contact items to each of the two project pages.
